### PR TITLE
Disallow purchase of listing by its seller

### PR DESCRIPTION
**Pull request by cuongdo**
### Checklist:

- [x] Code contains relevant tests for the problem you are solving
- [x] Ensure all new and existing tests pass
- [x] Update any relevant READMEs and [docs](https://github.com/OriginProtocol/docs)
- [x] Submit to the `develop` branch instead of `master`

### Description:

* Remove purchasing account's own listings in tests
  * This is a prerequisite for the following change
* Forbid sellers from purchasing their own listing

As is probably obvious, I'm inexperienced with modern JS, so pointers on how I can be more idiomatic and other tips would be appreciated.

### CODE REVIEW

**Comment by micahalcorn** [comment]
Thanks for doing this, @cuongdo 🙏

File: src/services/contract-service.js [hunk 1]
```diff
@@ -74,12 +75,33 @@ class ContractService {
...
+    if (this.currentAccountOverride !== undefined) {
+      return this.currentAccountOverride
+    }
     const accounts = await this.web3.eth.getAccounts()
     return accounts[0]
   }
 
+  // Runs the given function (which takes no parameters) as the provided
+  // account.
+  async asAccount(account, fn) {
```
[the remark this answers is not in the document]
Comment (cuongdo) [reply src/services/contract-service.js]: Yes, I'm not entirely happy with this either. One short-term compromise is to name the method to make it clear that it's only for testing. e.g. `asAccountForTestingOnly`.

That said, I'd definitely welcome something cleaner.

[the remark this answers is not in the document]
Comment (DanielVF) [reply src/services/contract-service.js]: Short version: I agree that just having an `opts={}` on each of our API methods that is passed through to our contract function calling code and can set the `from` address is the best possible solution.

----

Some kind of of "temporary state and guaranteed cleanup" is pretty common in the world.

Python:

    with open("welcome.txt") as file:
        data = file.read()

Ruby:

    File.open('Leo Tolstoy - War and Peace.txt', 'r') do |f|
        data = f.read
    end

or C#:

    using (var myFile = File.Create(myPath))
    {
       throw new Exception("Should not be able to create file")
    }

Javascript doesn't have built in support for this concept. But most implementations of it that I've seen just use callbacks (with async support). For example, all our unit tests work like this:

    it("should be cool!", async()=>{ 
       fail()
    })

The non-callback version is susceptible the same possible state issues, and adds the possibility of forgetting to reset it.

    contractService.setAccount(seller)
    await listings.sellerGetPayout()
    contractService.useDefaultAccount()

This just seems cleaner.

    contractService.asAccount(seller, async () => {
        await listings.sellerGetPayout()
    }

The downside though is that this is vulnerable to bugs if multiple copies, some with asAccount and some without are run in parallel. We shouldn't release this to the world.

We could also have the this asAccount method be on the origin object, and have it return a new origin object that uses the desired account to the async callback.

    origin.asAccount(seller, async (o) => {
        await o.listings.sellerGetPayout()
    }

HOWEVER, I do agree that it's cleanest, and most closely follows our functional style to just set an option on the api call.

    listings.sellerGetPayout({from:seller})

Glad you guys have my back when I propose things that have better options out there.

[the remark this answers is not in the document]
Comment (tyleryasaka) [reply src/services/contract-service.js]: Good points @DanielVF  - The purist in me definitely prefers the functional approach, but yes that is true that the set and reset pattern is quite common, as in the examples you listed. I'm more comfortable with this pattern if we do it in a clean way - monkey-patching our contract service for the purpose of testing just doesn't sit well with me.

I've created a PR with a similar but (imo) cleaner approach: https://github.com/OriginProtocol/origin-js/pull/223

See what you think, @DanielVF @cuongdo ?

**Comment by tyleryasaka** [comment]
Closing as https://github.com/OriginProtocol/origin-js/pull/223 has been merged in its place

**Pull request closed without merging**